File: Week 3/java/16.Check_no_Sequence.java
/*
 * WEEK 3 - LOOPS & NUMBER THEORY
 * Topic: Check Number Sequence (Decreasing then Increasing)
 * File: 16.Check_no_Sequence.java
 *
 * CONCEPT:
 * Given a sequence of N integers, determine if it can be split into two parts:
 * - First part: strictly decreasing
 * - Second part: strictly increasing
 * The transition from decreasing to increasing can happen at most once.
 * A fully increasing or fully decreasing sequence is also valid.
 * Equal consecutive elements make the sequence invalid.
 *
 * KEY POINTS / ALGORITHM:
 * 1. Read N and the first number as 'prev'.
 * 2. Set isDec = true (assume sequence starts in decreasing phase).
 * 3. For each subsequent number:
 *    a. If current == prev: print "false" (not strictly monotonic).
 *    b. If current < prev (decreasing):
 *       - If isDec is false (was already increasing), print "false" (can't decrease again).
 *       - Otherwise, continue (still in decreasing phase).
 *    c. If current > prev (increasing):
 *       - If isDec is true, set isDec = false (transition to increasing phase).
 *       - Otherwise, continue (still in increasing phase).
 *    d. Update prev = current.
 * 4. If loop completes without returning false, print "true".
 *
 * Time Complexity: O(n) - single pass through the sequence
 * Space Complexity: O(1) - only stores previous value and a boolean flag
 *
 * Input Format:
 *   Line 1: Integer N (number of elements)
 *   Lines 2..N+1: One integer per line
 * Output Format: "true" or "false"
 *
 * Examples:
 *   [5, 3, 1, 2, 4] -> true (decreasing 5,3,1 then increasing 1,2,4)
 *   [1, 2, 3, 4, 5] -> true (fully increasing, decreasing part is empty)
 *   [5, 4, 3, 2, 1] -> true (fully decreasing, increasing part is empty)
 *   [1, 2, 3, 2, 1] -> false (increases then decreases - invalid)
 *   [1, 2, 2, 3]    -> false (equal elements - not strictly increasing)
 */



import java.util.Scanner;
public class Main {
     public static void main(String[] args) {
          Scanner s = new Scanner(System.in);
          int n = s.nextInt();           // number of elements in the sequence
          int prev = s.nextInt();        // read first element
          int count = 2, current;
          boolean isDec = true;          // flag: true = in decreasing phase

          // Process remaining elements one by one
          while(count <= n) {
              current = s.nextInt();     // read next element
              count++;

              // Equal consecutive elements -> not strictly monotonic
              if(current == prev) {
                  System.out.println("false");
                    return;
                  }

              // Current is less than previous -> sequence is decreasing here
              if(current < prev) {
                    if(isDec == false) {  // was in increasing phase, can't decrease again
                        System.out.println("false");
                        return;
             }
              }
              // Current is greater than previous -> sequence is increasing here
                     else {
                        if(isDec == true) {  // transition from decreasing to increasing
                            isDec = false;   // mark that we've entered increasing phase
                }
           }
                 prev = current;         // update prev for next comparison
       }

          // If we made it through the entire sequence, it's valid
          System.out.println("true");
}
}

File: Week 3/cpp/16.check_no_sequence.cpp
/*
 * WEEK 3 - C++ LOOPS & NUMBER THEORY
 * Topic: Decreasing-then-Increasing Sequence Check
 * File: 16.check_no_sequence.cpp
 *
 * PROBLEM:
 *  N integers; return true if the sequence is strictly decreasing then
 *  strictly increasing (transition allowed at most once). Equal consecutive
 *  values -> false.
 *
 * KEY POINTS:
 *  - Track current phase with a bool `is_dec`.
 *  - Equal consecutive => false immediately.
 *  - Once we transition to increasing, decreasing again => false.
 *
 * COMPLEXITY: O(n) time, O(1) space.
 */

#include <iostream>
#include <vector>

bool check(const std::vector<int>& s) {
    if (s.size() < 2) return true;
    bool is_dec = true;
    for (size_t i = 1; i < s.size(); ++i) {
        if (s[i] == s[i - 1]) return false;
        if (s[i] < s[i - 1]) {
            if (!is_dec) return false;       // already increasing -> can't decrease
        } else {
            if (is_dec) is_dec = false;       // transition once
        }
    }
    return true;
}

int main() {
    int n;
    if (!(std::cin >> n)) {
        std::cout << std::boolalpha;
        for (auto v : std::vector<std::vector<int>> {
                 {5, 3, 1, 2, 4}, {1, 2, 3, 4, 5}, {5, 4, 3, 2, 1},
                 {1, 2, 3, 2, 1}, {1, 2, 2, 3} }) {
            std::cout << "{ ";
            for (int x : v) std::cout << x << ' ';
            std::cout << "} -> " << check(v) << '\n';
        }
        return 0;
    }
    std::vector<int> s(n);
    for (auto& x : s) std::cin >> x;
    std::cout << std::boolalpha << check(s) << '\n';
    return 0;
}

/*
 * NOTES:
 *  - Linear scan with a single boolean state -- clean and fast.
 *  - For STREAMED input you don't need to store the whole sequence; just
 *    keep `prev` and `is_dec`.
 */
